File: src/core/execution/nodes/scheduler.py
"""
Scheduler Node
Autonomous node that triggers connected nodes at random intervals
"""
import random
import time
from typing import Dict, Any, Optional

from ..node_base import BaseNode, ExecutionContext, ExecutionMode
from ....utils.logger import get_logger

logger = get_logger(__name__)


class SchedulerNode(BaseNode):
    """
    Scheduler node for autonomous workflow execution
    
    Fires a trigger signal at random intervals between min_seconds and max_seconds.
    Connected nodes will execute when the trigger fires.
    
    Properties (from metadata):
        min_seconds: Minimum interval between triggers (default: 5)
        max_seconds: Maximum interval between triggers (default: 10)
        enabled: Whether the scheduler is active (default: True)
    
    Outputs:
        trigger: Boolean signal (True when trigger fires)
        tick_count: Number of times the scheduler has fired
        timestamp: Unix timestamp of the trigger
    """
    
    # This is a CONTINUOUS node - it runs on every tick
    execution_mode = ExecutionMode.CONTINUOUS
    
    def __init__(self, node_id: str, node_data: Dict[str, Any]):
        super().__init__(node_id, node_data)
        
        # Get configuration from metadata
        self._min_seconds = float(self.metadata.get('min_seconds', 5.0))
        self._max_seconds = float(self.metadata.get('max_seconds', 10.0))
        self._enabled = self.metadata.get('enabled', True)
        
        # Validate
        if self._min_seconds > self._max_seconds:
            self._min_seconds, self._max_seconds = self._max_seconds, self._min_seconds
        
        # State
        self._last_fire_time: float = 0.0
        self._next_interval: float = self._generate_interval()
        self._fire_count: int = 0
        
        logger.debug(
            f"[Scheduler {node_id}] Initialized: "
            f"interval={self._min_seconds}-{self._max_seconds}s, "
            f"enabled={self._enabled}"
        )
    
    def _generate_interval(self) -> float:
        """Generate a random interval between min and max seconds"""
        return random.uniform(self._min_seconds, self._max_seconds)
    
    def execute(self, context: ExecutionContext) -> Dict[str, Any]:
        """
        Execute scheduler node (called once at workflow start)
        
        For continuous execution, use on_tick() instead.
        """
        # Initialize last fire time to now so we don't immediately fire
        self._last_fire_time = time.time()
        self._next_interval = self._generate_interval()
        
        return {
            'trigger': False,
            'tick_count': self._fire_count,
            'timestamp': time.time(),
            'next_fire_in': self._next_interval
        }
    
    def on_tick(self, context: ExecutionContext) -> Optional[Dict[str, Any]]:
        """
        Called on each tick to check if scheduler should fire
        
        Args:
            context: Execution context
            
        Returns:
            Output dict with trigger=True if firing, None otherwise
        """
        if not self._enabled:
            return None
        
        current_time = time.time()
        elapsed = current_time - self._last_fire_time
        
        # Check if interval has elapsed
        if elapsed >= self._next_interval:
            self._fire_count += 1
            self._last_fire_time = current_time
            self._next_interval = self._generate_interval()
            
            logger.info(
                f"[Scheduler {self.node_id}] Fired! "
                f"count={self._fire_count}, "
                f"next_in={self._next_interval:.2f}s"
            )
            
            return {
                'trigger': True,
                'tick_count': self._fire_count,
                'timestamp': current_time,
                'next_fire_in': self._next_interval
            }
        
        # Not time to fire yet
        return None
    
    def set_enabled(self, enabled: bool) -> None:
        """Enable or disable the scheduler"""
        self._enabled = enabled
        logger.info(f"[Scheduler {self.node_id}] {'Enabled' if enabled else 'Disabled'}")
    
    def reset(self) -> None:
        """Reset the scheduler state"""
        self._last_fire_time = time.time()
        self._next_interval = self._generate_interval()
        self._fire_count = 0
        logger.info(f"[Scheduler {self.node_id}] Reset")
    
    def get_status(self) -> Dict[str, Any]:
        """Get current scheduler status"""
        current_time = time.time()
        elapsed = current_time - self._last_fire_time
        time_until_next = max(0, self._next_interval - elapsed)
        
        return {
            'enabled': self._enabled,
            'fire_count': self._fire_count,
            'min_seconds': self._min_seconds,
            'max_seconds': self._max_seconds,
            'time_until_next': time_until_next,
            'last_fire_time': self._last_fire_time
        }

File: ts/src/core/execution/nodes/scheduler.ts
/**
 * SchedulerNode – autonomous node that fires a trigger at random intervals.
 * Mirrors Python src/core/execution/nodes/scheduler.py
 *
 * This is a CONTINUOUS node. The WorkflowRunner calls onTick() every ~100ms;
 * the node keeps its own timer and only fires when the random interval elapses.
 *
 * Instance state (lastFireTime, nextInterval, fireCount) lives on the node
 * object itself, which the runner keeps alive across ticks.
 */
import { BaseNode, ExecutionContext, ExecutionMode } from "../nodeBase";
import { WorkflowData, NodeID } from "../../types";
import { getLogger } from "../../../utils/logger";

const logger = getLogger("scheduler");

export class SchedulerNode extends BaseNode {
  // ── CONTINUOUS execution mode ──────────────────────────────────────
  static override executionMode = ExecutionMode.CONTINUOUS;

  // ── Instance state ─────────────────────────────────────────────────
  private _minSeconds: number;
  private _maxSeconds: number;
  private _enabled: boolean;
  private _lastFireTime = 0;
  private _nextInterval: number;
  private _fireCount = 0;

  constructor(nodeId: string, nodeData: import("../../types").NodeData) {
    super(nodeId, nodeData);

    const meta = this.metadata;
    let minRaw = Number(meta.min_seconds ?? meta.interval_seconds ?? 60);
    let maxRaw = Number(meta.max_seconds ?? meta.interval_seconds ?? 60);
    this._enabled = meta.enabled !== false;

    // Sanitize: fall back to sane defaults for NaN / Infinity
    if (!Number.isFinite(minRaw)) minRaw = 60;
    if (!Number.isFinite(maxRaw)) maxRaw = 60;

    this._minSeconds = minRaw;
    this._maxSeconds = maxRaw;

    if (this._minSeconds > this._maxSeconds) {
      [this._minSeconds, this._maxSeconds] = [this._maxSeconds, this._minSeconds];
    }

    this._nextInterval = this._generateInterval();

    logger.debug(
      `[Scheduler ${nodeId}] Initialized: interval=${this._minSeconds}-${this._maxSeconds}s, enabled=${this._enabled}`
    );
  }

  private _generateInterval(): number {
    return (
      Math.random() * (this._maxSeconds - this._minSeconds) + this._minSeconds
    );
  }

  // ── initialize() — called once when runner starts the workflow ──────
  override initialize(
    _workflow: WorkflowData,
    _allNodes: Map<NodeID, BaseNode>
  ): void {
    this._lastFireTime = Date.now() / 1000;
    this._nextInterval = this._generateInterval();
    logger.debug(
      `[Scheduler ${this.nodeId}] Initialized: first fire in ${this._nextInterval.toFixed(2)}s`
    );
  }

  // ── execute() — returns current state (no side effects) ────────────
  execute(_context: ExecutionContext): Record<string, unknown> {
    return {
      trigger: false,
      tick_count: this._fireCount,
      timestamp: this._lastFireTime,
      next_fire_in: this._nextInterval,
    };
  }

  // ── onTick() — called every ~100ms by WorkflowRunner ──────────────
  onTick(_context: ExecutionContext): Record<string, unknown> | null {
    if (!this._enabled) return null;

    // Pick up wired input values (e.g. from a Text node with process.env)
    const minInput = this.getInputValue("min_seconds", _context, undefined);
    const maxInput = this.getInputValue("max_seconds", _context, undefined);
    if (minInput != null) {
      const v = Number(minInput);
      if (Number.isFinite(v) && v > 0 && v !== this._minSeconds) {
        this._minSeconds = v;
        logger.debug(`[Scheduler ${this.nodeId}] min_seconds updated to ${v}`);
      }
    }
    if (maxInput != null) {
      const v = Number(maxInput);
      if (Number.isFinite(v) && v > 0 && v !== this._maxSeconds) {
        this._maxSeconds = v;
        logger.debug(`[Scheduler ${this.nodeId}] max_seconds updated to ${v}`);
      }
    }

    const now = Date.now() / 1000;
    const elapsed = now - this._lastFireTime;

    if (elapsed >= this._nextInterval) {
      this._fireCount++;
      this._lastFireTime = now;
      this._nextInterval = this._generateInterval();

      logger.info(
        `[Scheduler ${this.nodeId}] Fired! count=${this._fireCount}, next_in=${this._nextInterval.toFixed(2)}s`
      );

      return {
        trigger: true,
        tick_count: this._fireCount,
        timestamp: now,
        next_fire_in: this._nextInterval,
      };
    }

    return null;
  }

  // ── Helpers used by the runner ─────────────────────────────────────

  /** Clear state so the scheduler fires immediately on next tick. */
  reset(): void {
    this._lastFireTime = 0;
    this._nextInterval = this._generateInterval();
    this._fireCount = 0;
    logger.debug(`[Scheduler ${this.nodeId}] Reset`);
  }

  /**
   * @deprecated Legacy static helpers — no-ops retained for backward compatibility.
   * State lives on the node instance now; use the instance reset() method instead.
   * These do not affect runtime state and will be removed in a future version.
   */
  private static _legacyFireTimes = new Map<string, number>();

  /** @deprecated No-op. Use instance reset() instead. */
  static resetWorkflow(workflowId: string): void {
    for (const key of SchedulerNode._legacyFireTimes.keys()) {
      if (key.startsWith(`${workflowId}:`)) {
        SchedulerNode._legacyFireTimes.delete(key);
      }
    }
  }

  /** @deprecated No-op. Use instance reset() instead. */
  static resetAll(): void {
    SchedulerNode._legacyFireTimes.clear();
  }
}
